samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java

```diff
@@ -84,6 +113,38 @@ public void stop() {
     }
   }
 
+  private boolean checkAndEstablishConnectionWithNewAM() {
```

Comment:
It might be nice to breakdown the check and the establishing part.
[1] It makes it easier to write simpler tests but still tests each logic separately
[2] allows handling and evolution of each modular functions. e.g we don't have any retry for establishing but only for checking today.

wdyt?

Reply:
modular does sound good. but in this case, the smaller functions are really small and dont accomplish much else than a single call to client or store. Hence keeping this same. Also, doesnt have an advantage with testing.